### PR TITLE
Move ServiceType to SDK

### PR DESCRIPTION
**Pull request by liadtal**
Move the ServiceType from "Metric" to SDK root, and optionally include it when creating resource ID. This will enable SDK users to generate the full resource ID, including the service type suffix

### CODE REVIEW

File: src/SDK/SmartDetectorsSDK/ResourceIdentifier.cs [hunk 1]
```diff
@@ -43,6 +43,19 @@ public struct ResourceIdentifier : IEquatable<ResourceIdentifier>
         private const string ResourceGroupRegexPattern = SubscriptionRegexPattern + "/resourceGroups/(?<resourceGroupName>[^/]*)";
         private const string ResourceRegexPattern = ResourceGroupRegexPattern + "/providers/(?<resourceProviderAndType>.*)/(?<resourceName>[^/]*)";
 
+        /// <summary>
+        /// A dictionary, mapping <see cref="ServiceType"/> enumeration values to matching presentation in URI
+        /// </summary>
+        private static readonly ReadOnlyDictionary<ServiceType, string> MapAzureServiceTypeToPresentationInUri =
```
Comment (EranGon) [inline src/SDK/SmartDetectorsSDK/ResourceIdentifier.cs]: >PresentationInUri [](start = 93, length = 17)

ResourceUriPath? 
not sure presentation has anything to do with it

**Comment by EranGon** [comment]
>     public enum ServiceType

will this have anything other the storage services?
if not - i think we should refactor a bit

---
Refers to: src/SDK/SmartDetectorsSDK/ServiceType.cs:16 in 5c48e7f. [](commit_id = 5c48e7f8ae8dc87b4902121f156214fd00ef26e2, deletion_comment = False)

File: src/SDK/SmartDetectorsSDK/ResourceIdentifier.cs (hunk 1 above)
Comment (EranGon) [reply src/SDK/SmartDetectorsSDK/ResourceIdentifier.cs]: or follow the pattern of MapResourceTypeToString and name it MapServiceTypeToString

---
In reply to: [281468910](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/63#discussion_r281468910) [](ancestors = 281468910)

**Comment by liadtal** [comment]
>     public enum ServiceType

I know of similar patterns in other resource types - SqlServer and VMSS (although the suffixes there are built differently compared to the storage account suffixes, so we will need to make some changes if we even want to support these types).

---
In reply to: [489918436](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/63#issuecomment-489918436) [](ancestors = 489918436)

---
Refers to: src/SDK/SmartDetectorsSDK/Metric/ServiceType.cs:16 in 5c48e7f. [](commit_id = 5c48e7f8ae8dc87b4902121f156214fd00ef26e2, deletion_comment = False)

**Comment by EranGon** [comment]
>     public enum ServiceType

so my suggestion:
a. rename this to StorageServiceType
b. add an extension method to ResourceIdentifier for ToResourceId(StorageServiceType serviceType) in this project (that will call the regular ToResourceId and concat the service path)

this way it will be easy to add other service types (new enum + extension method)

---
In reply to: [489923665](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/63#issuecomment-489923665) [](ancestors = 489923665,489918436)

---
Refers to: src/SDK/SmartDetectorsSDK/Metric/ServiceType.cs:16 in 5c48e7f. [](commit_id = 5c48e7f8ae8dc87b4902121f156214fd00ef26e2, deletion_comment = False)

**Comment by liadtal** [comment]
>     public enum ServiceType

Agree regarding the extension method.
I didn't rename ServiceType to StorageServiceType - that's a big breaking change (mainly because the MetricClient accepts it in all its methods), and I prefer not to risk it at this point.

---
In reply to: [489962916](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/63#issuecomment-489962916) [](ancestors = 489962916,489923665,489918436)

---
Refers to: src/SDK/SmartDetectorsSDK/Metric/ServiceType.cs:16 in 5c48e7f. [](commit_id = 5c48e7f8ae8dc87b4902121f156214fd00ef26e2, deletion_comment = False)

**Comment by EranGon** [comment]
>     public enum ServiceType

lets discuss F2F - it's not managable to have a single enum with all service types of all possible resource types

---
In reply to: [490001573](https://github.com/Azure/SmartDetectorsMonitoringAppliance/pull/63#issuecomment-490001573) [](ancestors = 490001573,489962916,489923665,489918436)

---
Refers to: src/SDK/SmartDetectorsSDK/Metric/ServiceType.cs:16 in 5c48e7f. [](commit_id = 5c48e7f8ae8dc87b4902121f156214fd00ef26e2, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/Clients/MetricClient.cs [hunk 2]
```diff
@@ -102,7 +102,7 @@ public async Task<IEnumerable<MetricDefinition>> GetResourceMetricDefinitionsAsy
         /// <returns>A <see cref="Task{TResult}"/> object that represents the asynchronous operation, returning the list of metric definitions</returns>
         public async Task<IEnumerable<MetricDefinition>> GetResourceMetricDefinitionsAsync(ResourceIdentifier resource, ServiceType azureResourceService, CancellationToken cancellationToken)
         {
-            string resourceFullUri = resource.GetResourceFullUri(azureResourceService);
+            string resourceFullUri = resource.ToResourceId(azureResourceService);
```
Comment (giladsu) [inline src/SDK/SmartDetectorsShared/Clients/MetricClient.cs]: >resourceFullUri [](start = 19, length = 15)

so this is just the resource id now? If so, rename this too...

File: src/SDK/SmartDetectorsSDK/ResourceIdentifierExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// <summary>
+    /// Extension methods for the <see cref="ResourceIdentifier"/> class
+    /// </summary>
+    public static class ResourceIdentifierExtensions
+    {
+        /// <summary>
+        /// A dictionary, mapping <see cref="StorageServiceType"/> enumeration values to matching resource ID suffix strings
+        /// </summary>
+        private static readonly ReadOnlyDictionary<StorageServiceType, string> MapServiceTypeToString =
```
Comment (EranGon) [inline src/SDK/SmartDetectorsSDK/ResourceIdentifierExtensions.cs]: >MapServiceTypeToString [](start = 79, length = 22)

MapStorageServiceTypeToString

**Pull request merged**